Add EIP-5131: ENS Subdomain Authentication (#5131)
ethereum/EIPs: EIPS/eip-5131.md

### Setting up one or many `authAddress` records
The pre-requisite assumes that the `mainAddress` has an ENS ETH resolver record configured.

1. Using your `mainAddress` wallet, sign into ens.domains and create a subdomain record for `mainENS` called `auth[0-9]*`.  This becomes the `authENS`
2. Set the ETH resolver record for the subdomain created in step 1 (`authENS`) to the `authAddress`.
3. Using `authAddress`, sign into ens.domains and set the ENS reverse record to `authENS`

Currently this EIP does not enforce an upper-bound on the number of `authAddress` entries you can include. Users can repeat this process with as many address as they like. 

### Authenticating `mainAddress` via `authAddress`
Control of `mainAddress` and ownership of `mainAddress` assets is proven if any one of associated `authAddress` is the msg.sender or has signed the message.

Practically, this would work by performing the following operations:
1. Get the reverse ENS record for `authAddress`
2. Parse `auth[0-9]*.<mainENS>` to determine the linked ENS
3. Do a lookup on the linked ENS record to determine the linked `mainAddress`

> Note that this specification allows for both contract level and client/server side validation of signatures.  It is not limited to smart contracts, which is why there is no proposed external interface definition.

Reviewer: ```
4. MUST get the reverse ENS record for `mainAddress` and verify that it matches `<mainENS>`.
```
Otherwise one could set up other ENS nodes (with auths) that point to `mainAddress` and authenticate via those.

Examples:
 - In order for you to edit your profile on OpenSea, you must sign a message with your wallet address.
 - In order to access NFT gated content, you must sign a message with the wallet containing the NFT.
 - In order to claim an airdrop, you must interact with the smart contract with the qualifying wallet address.
 - In order to prove ownership of an NFT, you must sign a payload with the address that owns that NFT.

Unfortunately, we've seen many cases where users have accidentally signed a malicious payload. The result is almost always a significant loss of assets associated with the signing address.

In addition to this, many users keep significant portions of their assets in 'cold storage'. With the increased security from 'cold storage' solutions, we usually see decreased accessibility because users naturally increase the barriers required to access these wallets.

> This EIP proposes a solution which uses the Ethereum Name Service Specification (EIP-137) as a way to link one or more signing wallets (presumably less secure) to authenticate a main wallet.

Reviewer 2: ```suggestion
This EIP proposes a solution which uses the Ethereum Name Service Specification (EIP-137) as a way to link one or more signing wallets (presumably less secure) to authenticate a main wallet for reversible operations.
```

I would specify reversible because "authentication" could mean e.g. signing transactions on behalf of the cold wallet, and you wouldn't want a hot wallet to be able to do that for obvious reasons. Specifying reversible means that it's clear that the changes made should be able to be undone.

From there, the benefits are twofold. This EIP gives users increased security via outsourcing potentially malicious signing operations to wallets that are more accessible (hot wallets), while being able to maintain the intended security assumptions of wallets that are not frequently used for signing operations.

## Motivation

> As explained in the abstract, the current method of using a person's main wallet for authentication of control and asset ownership is both insecure and inconvenient.  It is as if, in order to authenticate, you must give an application root access to your wallet.  This proposal provides an easy way to do 'read only' access to a wallet by leveraging existing infrastructure.

Reviewer 2: ```suggestion
As explained in the abstract, the current method of using a person's main wallet for authentication of control and asset ownership is both insecure and inconvenient. It is as if, in order to authenticate, you must give an application root access to your wallet. This proposal provides an easy way to do reversible access to a wallet by leveraging existing infrastructure.
```

> Some solutions propose dedicated registry smart contracts to create this link, or new protocols to be supported.  Rather than 're-invent the wheel', this proposal aims to use the widely adopted Ethereum Name Service in order to bootstrap a safer and more convenient way to sign and authenticate.

Reviewer 2: ```suggestion
Some solutions propose dedicated registry smart contracts to create this link, or new protocols to be supported. Rather than 're-invent the wheel', this proposal aims to use the widely adopted Ethereum Name Service in order to bootstrap a safer and more convenient way to sign and authenticate.
```

## Specification
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

Let:
 - `mainAddress` represent the wallet address we are trying to authenticate or very asset ownership for
 - `mainENS` represent the reverse lookup ENS string for `mainAddress`

>  - `authAddress` represent the address we want to sign with in lieu of `mainAddress`

Reviewer 3: Are we going to add any requirement on this being an EOA?

Reviewer 3: No I don't think so. On second thoughts, we shouldn't be restricting it because a smart contract address could be used pretty easily with the right setup.

 - `authENS` represent the reverse lookup ENS string for `authAddress`.  It must be in the format `auth[0-9]*.<mainENS>`.

### Setting up one or many `authAddress` records
The pre-requisite assumes that the `mainAddress` has an ENS ETH resolver record configured.

> 1. Using your `mainAddress` wallet, sign into ens.domains and create a subdomain record for `mainENS` called `auth[0-9]*`.  This becomes the `authENS`

Reviewer 2: ```suggestion
1. Using your `mainAddress` wallet, sign into ens.domains and create a subdomain record for `mainENS` called `auth[0-9]*`. This becomes the `authENS`
```

> 1. Using your `mainAddress` wallet, sign into ens.domains and create a subdomain record for `mainENS` called `auth[0-9]*`.  This becomes the `authENS`

Author: Didn't think there would be a need for an upper limit.

> 2. Set the ETH resolver record for this `auth[0-9]*` subdomain to the `authAddress`

Reviewer 3: ```
2. Set the ETH resolver record for the subdomain created in step 1 (`authENS`) to the `authAddress`.
```

3. Using `authAddress`, sign into ens.domains and set the ENS reverse record to `authENS`

> Repeat this process with as many addresses as you would like.

Reviewer 3: Suggested edit: 
```
Currently this EIP does not enforce an upper-bound on the number of `authAddress` entries you can include. Users can repeat this process as with as many address as they like. 
```

Further, I have multiple devices:
 - My iPhone has mobile metamask hot wallet, and I can add it as an auth account by setting auth1.wilkins.eth to that wallet (and the corresponding reverse record). 
 - My iPad has the same and can be set to auth2.wilkins.eth, etc. 

Lost my device? No problem, just delete the record.  No need to import/reimport seed phrases.

## Reference Implementation

### Client/Server Side
In typescript, the validation function, using ethers.js would be as follows:
```
async function getLinkedAddress(provider: ethers.providers.Provider, address: string): Promise<LinkedAddress | null> {
  const addressENS = await provider.lookupAddress(address);
  if (!addressENS) return null;

  const authMatch = addressENS.match(/^(auth[0-9]*)\.(.*)/);
  if (!authMatch) return null;

  const linkedENS = authMatch[2];
  const linkedAddress = await provider.resolveName(linkedENS);

  if (!linkedAddress) return null;

  return {
    ens: linkedENS,
    address: linkedAddress
  };
}
```

### Contract side

### With a secure server

> If there is a secure, you could run the client/server code above, then use the result in conjunction with specs like [EIP-1721](./eip-1721.md): `Standard Signature Validation Method for Contracts` for a cheap and secure way to validate that the the message signer is indeed authenticated for the main address.

Reviewer 3: `If there is a secure server,`

Reviewer 3: I think it's worth elaborating a little more on this point.

            namehash = keccak256(
              abi.encodePacked(namehash, keccak256(abi.encodePacked(_nameParts[i-1])))
            );
            unchecked {
              i--;
            }
        }
    }

    function _computeReverseNamehash() private view returns (bytes32 namehash) {
        namehash = 0x0000000000000000000000000000000000000000000000000000000000000000;
        namehash = keccak256(
          abi.encodePacked(namehash, keccak256(abi.encodePacked('reverse')))
        );
        namehash = keccak256(
          abi.encodePacked(namehash, keccak256(abi.encodePacked('addr')))
        );
        namehash = keccak256(
          abi.encodePacked(namehash, keccak256(_addressToStringLowercase(msg.sender)))
        );
    }

    function _addressToStringLowercase(address _address) private pure returns (bytes memory addressString) {
        addressString = new bytes(40);
        for (uint i = 0; i < 20; i++) {
            bytes1 b = bytes1(uint8(uint(uint160(_address)) / (2**(8*(19 - i)))));
            bytes1 hi = bytes1(uint8(b) / 16);
            bytes1 lo = bytes1(uint8(b) - 16 * uint8(hi));
            addressString[2*i] = _bytes1ToChar(hi);

>             addressString[2*i+1] = _bytes1ToChar(lo);            

Reviewer 2: ```suggestion
            addressString[2*i+1] = _bytes1ToChar(lo);
```

---
eip: 5131
title: ENS Subdomain Authentication

> description: Use ENS to link hot wallets as approved signers for a root ENS address.

Reviewer 3: Using ENS subdomains to facilitate safer and more convenient signing operations.

### Authenticating `mainAddress` via `authAddress`
Control of `mainAddress` and ownership of `mainAddress` assets is proven if any one of associated `authAddress` is the msg.sender or has signed the message.

Practically, this would work by performing the following operations:
1. Get the reverse ENS record for `authAddress`
2. Parse `auth[0-9]*.<mainENS>` to determine the linked ENS
3. Do a lookup on the linked ENS record to determine the linked `mainAddress`

Note that this specification allows for both contract level and client/server side validation of signatures.  It is not limited to smart contracts, which is why there is no proposed external interface definition.

## Rationale
The proposed specification allows one to link multiple addresses as 'authentication addresses' to a core main address. This is beneficial from a security standpoint (if the authentication address is compromised, the assets held by the main address is not), and convenience (if the authentication address is a simple MetaMask wallet but the main address is a hardware wallet).

> ### Example:

Reviewer 3: An event requires attendees to prove ownership of an NFT to gain access. The attendee in question only has their phone available to perform a signing operation upon the presentation of a challenge. The phone is considered the attendees 'hot wallet' which is linked to `auth.wilkins.eth`.
The NFT is owned by the Ethereum address that corresponds to `wilkins.eth`, which controlled by the attendees cold storage device (Ledger, Trezor etc). The attendee does not have this device present at the event. 
The attendee is able to prove ownership of the NFT by signing with the 'hot wallet' key. 

This is made possible because the attendee completed the necessary upfront authorization steps with the cold storage device to allow the 'hot wallet' to be used for this operation.

I need to own an NFT to get into an event. I have my phone (hot wallet linked to auth.wilkins.eth), but not my ledger (wilkins.eth) If I sign with my phone, this verifies I own the main wallet and the NFTs in the ledger, letting me into the event safely.

Further, I have multiple devices:

>  - My iPhone has mobile metamask hot wallet, and I can add it as an auth account by setting auth1.wilkins.eth to that wallet (and the corresponding reverse record). 

Reviewer 3: I'd steer clear of mentioning the types of phones, but that's just me. Device 1 and device 2 is more generic.

Lost Access to Signing Key/Device: 
In the event that a user losses access to the signing key of an `authAddress` they can simply delete the corresponding ENS record. An attractive side effect of this is that there is no need to import/reimport seed phrases. 
In the unfortunate event that you lose access to the signing key of the `mainAddress`, users will lose access to that top-level domain and all related subdomains. As always, this private key data should be treated as highly sensitive. 

## Reference Implementation

### Client/Server Side
In typescript, the validation function, using ethers.js would be as follows:
```
async function getLinkedAddress(provider: ethers.providers.Provider, address: string): Promise<LinkedAddress | null> {
  const addressENS = await provider.lookupAddress(address);
  if (!addressENS) return null;

  const authMatch = addressENS.match(/^(auth[0-9]*)\.(.*)/);
  if (!authMatch) return null;

  const linkedENS = authMatch[2];
  const linkedAddress = await provider.resolveName(linkedENS);

  if (!linkedAddress) return null;

  return {
    ens: linkedENS,
    address: linkedAddress
  };
}
```

### Contract side

### With a secure server

> If your application operates a secure backend server, you could run the client/server code above, then use the result in conjunction with specs like [EIP-1721](./eip-1721.md): `Standard Signature Validation Method for Contracts` for a cheap and secure way to validate that the the message signer is indeed authenticated for the main address.

Reviewer 3: typo: 

```
[EIP-1271](./eip-1271.md) 
```

        namehash = 0x0000000000000000000000000000000000000000000000000000000000000000;
        namehash = keccak256(
          abi.encodePacked(namehash, keccak256(abi.encodePacked('reverse')))
        );
        namehash = keccak256(
          abi.encodePacked(namehash, keccak256(abi.encodePacked('addr')))
        );
        namehash = keccak256(
          abi.encodePacked(namehash, keccak256(_addressToStringLowercase(msg.sender)))
        );
    }

    function _addressToStringLowercase(address _address) private pure returns (bytes memory addressString) {
        addressString = new bytes(40);
        for (uint i = 0; i < 20; i++) {
            bytes1 b = bytes1(uint8(uint(uint160(_address)) / (2**(8*(19 - i)))));
            bytes1 hi = bytes1(uint8(b) / 16);
            bytes1 lo = bytes1(uint8(b) - 16 * uint8(hi));
            addressString[2*i] = _bytes1ToChar(hi);
            addressString[2*i+1] = _bytes1ToChar(lo);            
        }
    }

    function _bytes1ToChar(bytes1 b) private pure returns (bytes1 c) {
        if (uint8(b) < 10) return bytes1(uint8(b) + 0x30);
        else return bytes1(uint8(b) + 0x57);
    }

}
```

## Security Considerations

> The core purpose of this EIP is to enhance security and promote a safer way to authenticate wallet control and asset ownership when the main wallet is not needed and assets held by the main wallet do not need to be moved.  Consider it a way to do 'read only' authentication.

Reviewer 3: I think that it's worth mentioning that when you register an `authAddress`, you need to be cognizant of the fact that this address can now be given access to various operations that it previously didn't have access to.

Reviewer 3: As per our convo: 

If i set an authAddress any application can now accept that in leu of the mainAddress so as a user I want to be okay with that fact. Otherwise we could introduce another concept that allows you to narrow the scope to applications.
e.g. `authOpenSea.<mainENS>`, `authManifold.<mainENS>`

If every third party app piggybacks on auth.<mainENS> then the user can inadvertently expose a single hot wallet key to more power than they intended.

Maybe `auth[a-zA-Z0-9]*.<mainENS>` would be useful?

Reviewer: I agree, I think `auth[a-zA-Z0-9]*` would be neat to have in general (and make the onchain pattern verification much simpler).

Reviewer 4: +1 for `auth[a-zA-Z0-9]*`, it will allow for more clarity as well when potentially managing multiple authAddresses for them to actually be named

Reviewer: interesting, but I'd argue that this is beyond the scope of this proposal. imo, it warrants an own eip since there are much more things to consider to standardise this.

anyways, I did not fully understand how this would work yet. would `mainENS` point to a standard ERC721 contract in this case? If so, it will be impossible to set the required reverse records. or did I understand something wrong here?

Reviewer 4: Yeah agree its out of scope, but this EIP just got me thinking 😅 

> If so, it will be impossible to set the required reverse records
Would think the `owner` of the ERC721 contract would be `mainENS` in this case

Anyways, maybe I'll throw up a discussion somewhere else

Reviewer: > Would think the owner of the ERC721 contract would be mainENS in this case

Yup, that could work. but it's outside the ERC721 specs. so it would only apply to a certain subset of collections that use the `Ownable` extension.
Definitely work discussing it though - please ping me if you spin something up.

This is made possible because the attendee completed the necessary upfront authorization steps with the cold storage device to allow the 'hot wallet' to be used for this operation. 

### Example 2: Multiple Devices
I have multiple devices and wish to sign with any of them for convenience.  In this situation:
 - My phone has mobile metamask hot wallet, and I can add it as an auth account by setting auth1.wilkins.eth to that wallet (and the corresponding reverse record). 
 - My tablet has the same and can be set to auth2.wilkins.eth, etc. 

Lost Access to Signing Key/Device: 
In the event that a user losses access to the signing key of an `authAddress` they can simply delete the corresponding ENS record. An attractive side effect of this is that there is no need to import/reimport seed phrases. 
In the unfortunate event that you lose access to the signing key of the `mainAddress`, users will lose access to that top-level domain and all related subdomains. As always, this private key data should be treated as highly sensitive. 

## Reference Implementation

### Client/Server Side
In typescript, the validation function, using ethers.js would be as follows:
```

> async function getLinkedAddress(provider: ethers.providers.Provider, address: string): Promise<LinkedAddress | null> {

Reviewer 3: Maybe worth including a reference to the interface `LinkedAddress`. 

```
export interface LinkedAddress {
  ens: string,
  address: string,
}
```

  const addressENS = await provider.lookupAddress(address);
  if (!addressENS) return null;

  const authMatch = addressENS.match(/^(auth[0-9]*)\.(.*)/);
  if (!authMatch) return null;

  const linkedENS = authMatch[2];
  const linkedAddress = await provider.resolveName(linkedENS);

  if (!linkedAddress) return null;

  return {
    ens: linkedENS,
    address: linkedAddress
  };
}
```

### Contract side

### With a secure server
If your application operates a secure backend server, you could run the client/server code above, then use the result in conjunction with specs like [EIP-1271](./eip-1271.md) : `Standard Signature Validation Method for Contracts` for a cheap and secure way to validate that the the message signer is indeed authenticated for the main address.

#### Without a secure server (web client only)
Provided is a refrence implementation for an internal function to verify that the message sender has an authentication link to the main address.

> ``````

Reviewer 3: can you change this to ```  as it is making the bottom sections be part of the code block.

Examples:
 - In order for you to edit your profile on OpenSea, you must sign a message with your wallet address.
 - In order to access NFT gated content, you must sign a message with the wallet containing the NFT.
 - In order to claim an airdrop, you must interact with the smart contract with the qualifying wallet address.
 - In order to prove ownership of an NFT, you must sign a payload with the address that owns that NFT.

Unfortunately, we've seen many cases where users have accidentally signed a malicious payload. The result is almost always a significant loss of assets associated with the signing address.

In addition to this, many users keep significant portions of their assets in 'cold storage'. With the increased security from 'cold storage' solutions, we usually see decreased accessibility because users naturally increase the barriers required to access these wallets.

This EIP proposes a solution which uses the Ethereum Name Service Specification (EIP-137) as a way to link one or more signing wallets (presumably less secure) to authenticate a main wallet.

> From there, the benefits are twofold. This EIP gives users increased security via outsourcing potentially malicious signing operations to wallets that are more accessible (hot wallets), while being able to maintain the intended security assumptions of wallets that are not frequently used for signing operations.

Reviewer 5: A lot of this sounds like motivation, rather than abstract.  The abstract should be a human readable description of the specification and should not include commentary on "why".

---
eip: 5131
title: ENS Subdomain Authentication
description: Using ENS subdomains to facilitate safer and more convenient signing operations.
author: Wilkins Chung (@wwhchung)
discussions-to: https://ethereum-magicians.org/t/eip-5131-ens-subdomain-authentication/9458
status: Draft
type: Standards Track
category: ERC
created: 2022-06-03
requires: 137

> --- 

Reviewer 6: ```suggestion
---
```

## Abstract
This EIP links one or more signing wallets via Ethereum Name Service Specification ([EIP-137](./eip-137.md)) to prove control and asset ownership of a main wallet.

## Motivation

> Proving ownership of an asset to a third party application in the Ethereum ecosystem is common. Users frequently sign payloads of data to authenticate themselves before gaining access to perform some operation.  However, this method--akin to giving the third party root access to one's main wallet--is both insecure and inconvenient.

Reviewer 6: ```suggestion
Proving ownership of an asset to a third party application in the Ethereum ecosystem is common. Users frequently sign payloads of data to authenticate themselves before gaining access to perform some operation. However, this method--akin to giving the third party root access to one's main wallet--is both insecure and inconvenient.
```

Some solutions propose dedicated registry smart contracts to create this link, or new protocols to be supported. This is problematic from an adoption standpoint, and there have not been any standards created for them. 

### Proposal: Use the Ethereum Name Service (EIP-137) ###
Rather than 're-invent the wheel', this proposal aims to use the widely adopted Ethereum Name Service in order to bootstrap a safer and more convenient way to sign and authenticate, and provide 'read only' access to a main wallet.

From there, the benefits are twofold. This EIP gives users increased security via outsourcing potentially malicious signing operations to wallets that are more accessible (hot wallets), while being able to maintain the intended security assumptions of wallets that are not frequently used for signing operations.

## Specification
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

Let:
 - `mainAddress` represents the wallet address we are trying to authenticate or prove asset ownership for.

>  - `mainENS` represents the reverse lookup ENS string for `mainAddress`.

Reviewer: I think it could be beneficial for the readers understanding to underline the importance of the reverse record here. Maybe adding something like
```
This is crucial because it acts as proof that `mainENS` is operated by or on behalf of the `mainAddress` owner.
```

 - `authAddress` represents the address we want to use for signing in lieu of `mainAddress`.
 - `authENS` represents the reverse lookup ENS string for `authAddress`.  It must be in the format `auth[0-9A-za-z]*.<mainENS>`.

### Setting up one or many `authAddress` records
The pre-requisite assumes that the `mainAddress` has an ENS ETH resolver record and reverse record configured.

> 1. Using your `mainAddress` wallet, sign into ens.domains and create a subdomain record for `mainENS` called `auth[0-9A-za-z]*`.  This becomes the `authENS`

Reviewer: Typo: `auth[0-9A-Za-z]*`

On second thought, would it make sense to completely open this? So that the label only has to begin with `auth`? The format of the label is anyways already pretty restricted by [UTS46](https://unicode.org/reports/tr46/) with the `transitional=false` and `useSTD3AsciiRules=true` as per [EIP-137](https://eips.ethereum.org/EIPS/eip-137).

Reviewer 3: I'm in favor of opening it up completely. My original concern was that `auth[0-9]*` was too restrictive. `auth[0-9A-Za-z]*` less so; anything more open is cool by me. Staying inline with EIP-137 seems pretty sensible.

> 1. Using your `mainAddress` wallet, sign into ens.domains and create a subdomain record for `mainENS` called `auth[0-9A-Za-z]*`.  This becomes the `authENS`

Reviewer 7: Oh, and please remove external links (`ens.domains`).

2. Set the ETH resolver record for the subdomain created in step 1 (`authENS`) to the `authAddress`.
3. Using `authAddress`, sign into ens.domains and set the ENS reverse record to `authENS`

Currently this EIP does not enforce an upper-bound on the number of `authAddress` entries you can include. Users can repeat this process with as many address as they like. 

### Authenticating `mainAddress` via `authAddress`
Control of `mainAddress` and ownership of `mainAddress` assets is proven if any one of associated `authAddress` is the msg.sender or has signed the message.

Practically, this would work by performing the following operations:
1. Get the reverse ENS record for `authAddress`

> 2. Parse `auth[0-9A-za-z]*.<mainENS>` to determine the linked ENS

Reviewer: Some type as above: `auth[0-9A-Za-z]*`

3. Do a lookup on the linked ENS record to determine the linked `mainAddress`
4. MUST get the reverse ENS record for `mainAddress` and verify that it matches `<mainENS>` 
    - Otherwise one could set up other ENS nodes (with auths) that point to `mainAddress` and authenticate via those.

Note that this specification allows for both contract level and client/server side validation of signatures.  It is not limited to smart contracts, which is why there is no proposed external interface definition.

> 

Reviewer 7: This reads more like a guide or instruction manual than a specification. I'd probably recommend reworking this section (sorry!) so that it describes what is valid/invalid, instead of instructions to get to a valid state.

---

For example:

```markdown
When delegating signing authority from `mainAddress` to `authAddress`:

The ENS entry for `mainENS` must have a subdomain record matching the format `auth[0-9A-Za-z]*`. This subdomain record is referred to as `authENS`.

The ENS entry for `authENS` must resolve to `authAddress`.
```

I've probably mixed up some of the terminology above, so I wouldn't recommend copying it verbatim :sweat_smile:

>     - Otherwise one could set up other ENS nodes (with auths) that point to mainAddress and authenticate via those.

Reviewer: format: put \` around `mainAddress`

## Rationale

> The proposed specification allows one to link multiple addresses as 'authentication addresses' to a core main address. This is beneficial from a security standpoint (if the authentication address is compromised, the assets held by the main address is not), and convenience (if the authentication address is a simple MetaMask wallet but the main address is a hardware wallet).

Reviewer 6: So I think this section is reading more like motivation than rationale. I think you should consider moving a fair bit of this content above and focus on answering why certain design decisions were chosen over other decisions here.

### Example 1: Event Access

> An event requires attendees to prove ownership of an NFT to gain access. The attendee in question only has their phone available to perform a signing operation upon the presentation of a challenge. The phone is considered the attendees 'hot wallet' which is linked to auth.wilkins.eth.

Reviewer 7: ```suggestion
An event requires attendees to prove ownership of an NFT to gain access. The attendee in question only has their phone available to perform a signing operation upon the presentation of a challenge. The phone is considered the attendee's 'hot wallet' which is linked to `auth.wilkins.eth`.
```

> The NFT is owned by the Ethereum address that corresponds to wilkins.eth, which is controlled by the attendees' cold storage device (Ledger, Trezor etc). The attendee does not have this device present at the event.

Reviewer 7: ```suggestion
The NFT is owned by the Ethereum address that corresponds to `wilkins.eth`, which is controlled by the attendee's cold storage device (Ledger, Trezor etc). The attendee does not have this device present at the event.
```

The attendee is able to prove ownership of the NFT by signing with the 'hot wallet' key.

This is made possible because the attendee completed the necessary upfront authorization steps with the cold storage device to allow the 'hot wallet' to be used for this operation. 

### Example 2: Multiple Devices

> I have multiple devices and wish to sign with any of them for convenience.  In this situation:

Reviewer 6: ```suggestion
I have multiple devices and wish to sign with any of them for convenience. In this situation:
```

 - My phone has mobile metamask hot wallet, and I can add it as an auth account by setting auth1.wilkins.eth to that wallet (and the corresponding reverse record). 
 - My tablet has the same and can be set to auth2.wilkins.eth, etc. 

Lost Access to Signing Key/Device: 
In the event that a user losses access to the signing key of an `authAddress` they can simply delete the corresponding ENS record. An attractive side effect of this is that there is no need to import/reimport seed phrases. 

> In the unfortunate event that you lose access to the signing key of the `mainAddress`, users will lose access to that top-level domain and all related subdomains. As always, this private key data should be treated as highly sensitive. 

Reviewer 7: This section still reads as Motivation, and not Rationale. Motivation should explain why the EIP needs to exist as a whole, while the Rationale should explain individual technical choices within the EIP itself. For example, your Rationale section could explain why you don't limit the number of `authAddress` entries (though I'm not saying you _have_ to explain that particular choice.)

## Reference Implementation

### Client/Server Side
In typescript, the validation function, using ethers.js would be as follows:
```
export interface LinkedAddress {
  ens: string,
  address: string,
}

async function getLinkedAddress(provider: ethers.providers.Provider, address: string): Promise<LinkedAddress | null> {
  const addressENS = await provider.lookupAddress(address);
  if (!addressENS) return null;

>   const authMatch = addressENS.match(/^(auth[0-9A-za-z]*)\.(.*)/);

Reviewer: Same typo: `auth[0-9A-Za-z]*`

  if (!authMatch) return null;

  const linkedENS = authMatch[2];
  const linkedAddress = await provider.resolveName(linkedENS);

  if (!linkedAddress) return null;

  return {
    ens: linkedENS,
    address: linkedAddress
  };
}
```

### Contract side

> ### With a secure server

Reviewer: Format: I think this should be `#### With a secure server` right?

If your application operates a secure backend server, you could run the client/server code above, then use the result in conjunction with specs like [EIP-1271](./eip-1271.md) : `Standard Signature Validation Method for Contracts` for a cheap and secure way to validate that the the message signer is indeed authenticated for the main address.

> #### Without a secure server (web client only)

Reviewer: I have trouble understanding what is meant by `secure server`. Does this refer to the node provider? Otherwise, I could run the above js code also on a web client.

            }
        }
    }

    function _computeNamehash(bytes32 parentNamehash, string calldata name)
        private
        pure
        returns (bytes32 namehash)
    {
        namehash = keccak256(abi.encodePacked(parentNamehash, keccak256(bytes(name))));
    }

    // _computeNamehash('addr.reverse')
    bytes32 constant ADDR_REVERSE_NODE =
        0x91d1777781884d03a6757a803996e38de2a42967fb37eeaca72729271025a9e2;

    function _computeReverseNamehash(address _address) private pure returns (bytes32 namehash) {
        namehash = keccak256(abi.encodePacked(ADDR_REVERSE_NODE, sha3HexAddress(_address)));
    }

    function sha3HexAddress(address addr) private pure returns (bytes32 ret) {
        assembly {
            let lookup := 0x3031323334353637383961626364656600000000000000000000000000000000
            let i := 40
            for {

            } gt(i, 0) {

            } {
                i := sub(i, 1)
                mstore8(i, byte(and(addr, 0xf), lookup))
                addr := div(addr, 0x10)
            }
            ret := keccak256(0, 40)
        }
    }
}
```

## Security Considerations

> The core purpose of this EIP is to enhance security and promote a safer way to authenticate wallet control and asset ownership when the main wallet is not needed and assets held by the main wallet do not need to be moved.  Consider it a way to do 'read only' authentication.

Reviewer 6: ```suggestion
The core purpose of this EIP is to enhance security and promote a safer way to authenticate wallet control and asset ownership when the main wallet is not needed and assets held by the main wallet do not need to be moved. Consider it a way to do 'read only' authentication.
```

---
eip: 5131
title: ENS Subdomain Authentication
description: Using ENS subdomains to facilitate safer and more convenient signing operations.
author: Wilkins Chung (@wwhchung)
discussions-to: https://ethereum-magicians.org/t/eip-5131-ens-subdomain-authentication/9458
status: Draft
type: Standards Track
category: ERC
created: 2022-06-03
requires: 137
---

## Abstract
This EIP links one or more signing wallets via Ethereum Name Service Specification ([EIP-137](./eip-137.md)) to prove control and asset ownership of a main wallet.

## Motivation
Proving ownership of an asset to a third party application in the Ethereum ecosystem is common. Users frequently sign payloads of data to authenticate themselves before gaining access to perform some operation. However, this method--akin to giving the third party root access to one's main wallet--is both insecure and inconvenient.

> *** Examples: ***

Reviewer 8: This EIP seems highly useful given these examples. At the moment there is a real pain point of holding assets securely while still being able to prove ownership via a hot wallet especially in mobile settings. I'd be excited to see this proposal go through

Practically, this would work by performing the following operations:
1. Get the reverse ENS record for `authAddress`
2. Parse `auth[0-9A-Za-z]*.<mainENS>` to determine the linked ENS
3. Do a lookup on the linked ENS record to determine the linked `mainAddress`
4. MUST get the reverse ENS record for `mainAddress` and verify that it matches `<mainENS>` 
    - Otherwise one could set up other ENS nodes (with auths) that point to `mainAddress` and authenticate via those.

Note that this specification allows for both contract level and client/server side validation of signatures.  It is not limited to smart contracts, which is why there is no proposed external interface definition.

## Rationale

### Usage of EIP-137
The proposed specification makes use of EIP-137 rather than introduce another registry paradigm. The reason for this is due to the existing wide adoption of EIP-137 and ENS.

However, the drawback to EIP-137 is that any linked `authAddress` must contain some ETH in order to set the `authENS` reverse record. This can be solved by a separate reverse lookup registry that enables `mainAddress` to set a reverse record with a message signed by `authAddress`.

> ### One-to-Many Authenitcation Relationship

Reviewer 2: ```suggestion
### One-to-Many Authentication Relationship
```

Further, you can design UX without any user interaction necessary to 'pick' the interacting address by display assets owned by `authAddress` and `mainAddress` and use the appropriate address dependent on the asset the user is attempting to authenticate with.

## Reference Implementation

### Client/Server Side
In typescript, the validation function, using ethers.js would be as follows:
```
export interface LinkedAddress {
  ens: string,
  address: string,
}

async function getLinkedAddress(provider: ethers.providers.Provider, address: string): Promise<LinkedAddress | null> {
  const addressENS = await provider.lookupAddress(address);
  if (!addressENS) return null;

  const authMatch = addressENS.match(/^(auth[0-9A-Za-z]*)\.(.*)/);
  if (!authMatch) return null;

  const linkedENS = authMatch[2];
  const linkedAddress = await provider.resolveName(linkedENS);

  if (!linkedAddress) return null;

  return {
    ens: linkedENS,
    address: linkedAddress
  };
}
```

### Contract side

> #### With a secure server
> If your application operates a secure backend server, you could run the client/server code above, then use the result in conjunction with specs like [EIP-1271](./eip-1271.md) : `Standard Signature Validation Method for Contracts` for a cheap and secure way to validate that the the message signer is indeed authenticated for the main address.
> 
> #### Without a secure server (web client only)

Reviewer 2: ```suggestion
#### With a backend
If your application operates a secure backend server, you could run the client/server code above, then use the result in conjunction with specs like [EIP-1271](./eip-1271.md) : `Standard Signature Validation Method for Contracts` for a cheap and secure way to validate that the the message signer is indeed authenticated for the main address.

#### Without a backend (JavaScript only)
```

function validate(address ensRegistry, bytes calldata senderENS, address mainAddress, string[] memory mainENSParts) view returns(bool);
```

> This can be an internal validate function to the smart contract, that is used in conjunction with specs like EIP-1721: Standard Signature Validation Method for Contracts and Merkle Proof Verification.

Reviewer 2: ```suggestion
This can be an internal validate function to the smart contract, that is used in conjunction with specs like [EIP-1721](./eip-1721.md): Standard Signature Validation Method for Contracts and Merkle Proof Verification.
```

Should be linked

Author: Thanks.  Have made edits.
